examples/cubature_recombination.py

```diff
@@ -0,0 +1,194 @@
+    return jnp.asarray(prod_nodes), jnp.prod(prod_weights_un_multiplied, axis=-1)
+
+
+def monomial_power_generator(
+    dimension: int, max_degree: int, *, mode: Literal["all", "even", "odd"] = "all"
+) -> Iterable[tuple[int, ...]]:
+    """
+    Return a generator for all combinations of multi-variate monomial powers.
+
+    :param dimension: Number of unique variates; dimension of the domain over which
+        the monomials are defined
+    :param max_degree: Maximal degree of any monomial power; equal to the sum of
+        powers for each variate in a given monomial E.G :math:`xy**2` has degree three
+        :math:`xy` has degree two, :math:`x` has degree one, etc...
+    :param mode: If to return 'all', 'even', or 'odd' multi-variate monomial powers; a
+        monomial is even if and only if its powers sum to a non-zero even value. E.G.
+        :math:`xy` is even, :math:`xy**2` is odd, and :math:`1 == x^0 y^0` is odd.
+    :return: generator for all multi-variate monomial powers of the specified dimension
+        and maximum degree.
+    """
+    monomial_degree_generator = itertools.product(
+        range(max_degree + 1),
+        repeat=dimension,
+    )
+    exact_monomial_degree_generator = itertools.filterfalse(
+        lambda x: sum(x) > max_degree,
+        monomial_degree_generator,
+    )
+    if mode == "all":
+        return exact_monomial_degree_generator
+    if mode == "even":
+        return itertools.filterfalse(
+            lambda x: sum(x) == 0 or sum(x) % 2 != 0,
+            exact_monomial_degree_generator,
+        )
+    if mode == "odd":
+        return itertools.filterfalse(
+            lambda x: sum(x) % 2 == 0, exact_monomial_degree_generator
+        )
+    raise ValueError("Invalid mode; must be one of ['all', 'even', 'odd'].")
+
+
+def test_functions_from_monomial_powers(
+    monomial_powers: Int[Array, "k n"],
+) -> Callable[[Float[Array, " n"]], Float[Array, " k"]]:
+    """
+    Construct test functions given a set of multi-variate monomial powers.
+
+    :param monomial_powers:
+    :return:
+    """
+    _, n = monomial_powers.shape
+    coefficients = jnp.zeros((n, n))
+    column_index = jnp.arange(n)
+
+    @jax.vmap
+    def reversed_coefficients(_power: Int[Array, " n"]) -> Int[Array, "n n"]:
+        """Create a polyval coefficient matrix given a multi-variate monomial power."""
+        return coefficients.at[_power, column_index].set(1)
+
+    @ft.partial(jax.vmap, in_axes=(0, None))
+    def monomial_test_function(
+        _coefficients: Int[Array, "n n"], x: Float[Array, " n"]
+    ) -> Float[Array, " "]:
+        """Evaluate a multi-variate monomial."""
+        return jnp.prod(jnp.polyval(_coefficients[::-1], x))
+
+    return jtu.Partial(monomial_test_function, reversed_coefficients(monomial_powers))
+
+
+def main(dimension: int = 3, max_degree: int = 4) -> Coresubset:
+    """
+    Run the 'recombination' example for finding non-product cubature formulae.
+
+    Generates the
```

Comment:
I think this has not been filled in before committing but should be

Reply:
Can you add some mathematical background on the problem we are trying to solve here when updating?